packages/client/internals/MonacoOutput.vue

```diff
@@ -0,0 +1,143 @@
+<script setup lang="ts">
+import { debounce } from '@antfu/utils'
+import { useVModel } from '@vueuse/core'
+import { computed, ref, watch } from 'vue'
+import { isDark } from '../logic/dark'
+import { isPrintMode } from '../logic/nav'
+import { runJavaScript } from '../logic/runCode'
+import { useSlideContext } from '../context'
+import IconButton from './IconButton.vue'
+
+const props = defineProps<{
+  modelValue: string
+  lang: string
+  autorun?: boolean | 'once'
+  height?: string
+}>()
+const emit = defineEmits(['update:modelValue'])
+const code = useVModel(props, 'modelValue', emit)
+
+const { $renderContext } = useSlideContext()
+const disabled = computed(() => !['slide', 'presenter'].includes($renderContext.value))
+
+const autorun = isPrintMode.value ? 'once' : props.autorun
+const output = ref(autorun ? '_running' : '_empty')
+
+let shikiModule: typeof import('#slidev/shiki') | undefined
+let tsModule: typeof import('typescript') | undefined
+
+const run = debounce(200, async () => {
+  if (disabled.value)
+    return
+
+  const setAsRunning = setTimeout(() => {
+    output.value = '_running'
+  }, 500)
+
+  const { shiki, themes } = (shikiModule ??= await import('#slidev/shiki'))
+  const highlighter = await shiki
+  const highlight = (code: string) => highlighter.codeToHtml(code, {
+    lang: 'javascript',
+    theme: typeof themes === 'string'
+      ? themes
+      : isDark.value
+        ? themes.dark || 'vitesse-dark'
+        : themes.light || 'vitesse-light',
+  })
+
+  const { transpile } = (tsModule ??= await import('typescript'))
+  const js = props.lang === 'typescript'
+    ? transpile(code.value, {
+      module: tsModule.ModuleKind.ESNext,
+    })
+    : code.value
+
+  output.value = (await runJavaScript(js)).flatMap(
+    ([type, content]) => [
+      `<div class="output-line">`,
+      `<span class="decorator">[</span>`,
+      `<span class="log-type ${type}">${type}</span>`,
+      `<span class="decorator">]:&nbsp;</span>`,
+      `<div class="select-text">`,
+      content.map(highlight).join('<span class="decorator">, </span>'),
+      `</div>`,
+      `</div>`,
+    ],
+  ).join('')
+
+  clearTimeout(setAsRunning)
+})
+
+if (autorun === 'once')
+  run()
+else if (autorun)
+  watch(code, run, { immediate: true })
+</script>
+
+<template>
+  <div
+    class="relative flex flex-col px-2 py-1 rounded-b bg-$slidev-code-background"
+    :style="{ height: props.height }"
+    data-waitfor=".output"
+  >
+    <div v-if="disabled" class="text-sm text-center opacity-50">
+      Code is disabled in the "{{ $renderContext }}" mode
+    </div>
+    <div v-else-if="output === '_empty'" class="text-sm text-center opacity-50">
+      Click the play button to run the code
+    </div>
+    <div v-else-if="output === '_running'" class="text-sm text-center opacity-50 running">
+      Running...
+    </div>
+    <template v-else>
+      <div class="mb-1 -mt-1 text-xs font-bold text-primary">
+        OUTPUT
+      </div>
```

Comment:
Made the change to use Vue to render instead of assign everything in HTML, this should give us better control for the styling, and also interactivity if needed in the future.

I also think we could have a `raw` module or so, where we prints the output plain text (without `LOG` indicator and so), for example I'd expect `console.log('Hello World')` should just printing plain `Hello World` instead of `LOG: "Hello World"`.

Reply:
Inspired by your advice, I am now wondering if `LOG` indicator is actualy needed. Most of the code people want to run in slides may only use `console.log`.

Reply:
Also, now `console.log('Hello')` will just output `Hello`, while `console.log(['Hello'])` will output `["Hello"]`.

Reply:
Removed log type indicator and `rawMode` option in b3792508929a26b32540fb0cdb7e93280ede1db0